### ISSUE TITLE
IPOPT : Optimization flag being ignored

### ISSUE THREAD

**Issue by junkudo** [spack/spack#6732]
My command line optimization flag is being over-written by the IPOPT build.
Spack command:
```
spack -d install -v ipopt@3.12.7%intel@16.0.3 cflags="-g -O0" cppflags="-g -O0" cxxflags="-g -O0" fflags="-g -O0" -coinhsl+metis arch=linux-rhel7-x86_64
```

Spack debug output snippet:
```
[cc] /usr/tce/packages/intel/intel-16.0.3/bin/icpc -I/g/g14/kudo4/spack/opt/spack/linux-rhel7-x86_64/intel-16.0.3/metis-4.0.3-sacetyhy56rhqu6ggx3hzjfcwsaxs5t3/include -I/g/g14/kudo4/spack/opt/spack/linux-rhel7-x86_64/intel-16.0.3/mumps-\
5.1.1-5a4editxji4yi2zavi6gwtism4mgjl5c/include -g -O0 -g -O0 -DHAVE_CONFIG_H -I. -I. -O3 -ip -mp1 -DNDEBUG -DIPOPT_BUILD -MT IpDebug.lo -MD -MP -MF .deps/IpDebug.Tpo -c IpDebug.cpp -fPIC -DPIC -o .libs/IpDebug.o

```
Note the `-O3` being introduced.

### Information on your system
Platform : TOSS3/TOSS (Tri-Lab Operating System Stack)

Relevant configuration details:
compilers.yaml
```
compilers:
- compiler:
    environment: {}
    extra_rpaths: []
    modules: []
    operating_system: rhel7
    paths:
      cc: /usr/tce/packages/intel/intel-16.0.3/bin/icc
      cxx: /usr/tce/packages/intel/intel-16.0.3/bin/icpc
      f77: /usr/tce/packages/intel/intel-16.0.3/bin/ifort
      fc: /usr/tce/packages/intel/intel-16.0.3/bin/ifort
    flags: {}
    spec: intel@16.0.3
    target: x86_64
```

packages.yaml:
```
packages:
  all:
    compiler: [intel, gcc, pgi, clang, xl, nag]
    providers:
      mpi: [mvapich2]
      blas: [netlib-lapack]
      lapack: [netlib-lapack]
  metis:
    version: [4.0.3]
  netlib-lapack:
    paths:
      netlib-lapack@3.4.2: /usr/lib64/
    buildable: False
  mvapich2:
    paths:
      mvapich2@2.3: /usr/tce/packages/mvapich2/mvapich2-2.3-intel-16.0.3/
    buildable: False
  mpich:
    buildable: False
  openmpi:
    buildable: False
  intel-mpi:
    buildable: False
  spectrum-mpi:
    buildable: False
  intel-parallel-studio:
    buildable: False
  cmake:
    paths:
      cmake@3.5.2 arch=linux-rhel7-x86_64: /usr/tce/packages/cmake/cmake-3.5.2/
    buildable: False
  pkg-config:
    paths:
      pkg-config@0.27.1 arch=linux-rhel7-x86_64: /usr/bin/
    buildable: False
  m4:
    paths:
      m4@1.4.16 arch=linux-rhe17-x86_64: /usr/bin/
    buildable: False
```

**Comment by adamjstewart** [comment]
See my comment [here](https://github.com/spack/spack/issues/6769#issuecomment-354323678). It's not that your flags are being ignored, it's that they are being handled by the compiler wrapper, not `CFLAGS` and friends.

**Comment by junkudo** [comment]
Hey.  Thanks for the reply. 
Since I'm looking at the debug output provided by the `-d` flag, I'm fairly certain I'm looking at the final call to the base compiler (so after the compiler wrapper injects the flags).

**Comment by adamjstewart** [comment]
Ah, sorry, I didn't notice that. So the problem isn't that `-O0` is being ignored, it's that it's being overridden? I just noticed #6640. Is this a duplicate of that issue?

I'll reopen this issue. Feel free to reopen the other issues that I closed. Or we can handle the followup discussion in #6640 since they all likely stem from the same thing.

**Comment by junkudo** [comment]
In this case, the optimization flag is being overridden by a default.  I think that's true of #6745 as well.  I don't know what's going on in #6749 .

When I posted #6640, I didn't know about the compiler wrappers so some of those packages I posted are correct.  I think the three issues I created are actual issues so I'll likely open up the three independently.  Relevant to this issue is PR #6714 which  I haven't gotten around to implementing.

**Comment by adamjstewart** [comment]
Thanks for all the links! #6714 seems like a working hack, but a more general solution would definitely be preferable. I don't know as much about the compiler wrappers as other developers so I'll let someone else try to tackle this one.

**Comment by junkudo** [comment]
Fixed with PR #6714
